WebGPURenderer: implement ClippingGroup object

aardgoose opened the pull request:
Fixed #28838.

As suggested in original renderer https://github.com/mrdoob/three.js/blob/aedb9c75bedff3a5f8476dbb8b41dd40c4b09603/src/renderers/WebGLRenderer.js#L1777

Adds support for ClippingGroup objects. Clipping planes are additive with nested groups as demonstrated in example.

Commit:
clipping groups

Replace renderer and material clipping planes with nestable clipping
groups

Mugen87 commented:
When I read the code correctly, this PR removes `Renderer.clippingPlanes`.

Is it also possible to remove the support for `Material.clippingPlanes`? When I understand the implementation correctly, `ClippingGroup` can handle both local and global clipping.

aardgoose commented:
@Mugen87 I have already removed material clipping planes in my copy.  I'll push it here soon. It gets simplifies the Renderer somewhat.

aardgoose commented:
@Mugen87 @sunag 

ClippingGroup() can now replicate the functionality of the existing API without the problems encountered with postprocessing passes etc #28838, the existing example has been updated to use this. The only clipping property now on the material object is 'alphaToCoverage' which is common to other materials.

the planes used for union testing and intersection testing are accumulated separately when inherited.


@sunag To retain the current behaviour where changing 'alphaToCoverage' results in a material rebuild, I have added boolean properties to the RenderObject.getMaterialCacheKey(), I don't know if this is the correct solution here?

Commit:
update to upstream

sunag commented:
> To retain the current behaviour where changing 'alphaToCoverage' results in a material rebuild, I have added boolean properties to the RenderObject.getMaterialCacheKey(), I don't know if this is the correct solution here?

I'll check this weekend, sorry for the delay.

aardgoose commented:
> > To retain the current behaviour where changing 'alphaToCoverage' results in a material rebuild, I have added boolean properties to the RenderObject.getMaterialCacheKey(), I don't know if this is the correct solution here?
> 
> I'll check this weekend, sorry for the delay.

The question is redundant now, fixed by a upstream PR from @Mugen87

Commit:
rework cache key

Commit:
Merge branch 'dev' into clipping-context-group

Commit:
use right cache key

Commit:
Merge branch 'dev' into clipping-context-group

Mugen87 commented:
@aardgoose Do you have the resources to rebase the PR? If not, I'm going to give it a try in the upcoming days.

We really should merge this in `r171` to eventually fix global clipping planes.

aardgoose commented:
Yes, I'll take a look, then I can revisit hardware clipping which is apparently coming in WebGPU in the next chrome version.

Mugen87 commented:
Great! Can't wait to have access to `ClippingGroup`. It makes things so much nicer.

Commit:
Merge branch 'dev' of https://github.com/mrdoob/three.js into clipping-context-group

Commit:
fixup

Commit:
fix cache key use

Commit:
clean up TSL

Commit:
Merge branch 'dev' into clipping-context-group

Commit:
remove unused import

Mugen87 committed:
Update ToonOutlinePassNode.js

Add missing `clippingContext`.

Mugen87 committed:
Update Renderer.js

Add default parameter to renderObject().

Mugen87 commented on src/renderers/common/Renderer.js:
```diff
@@ -1584,23 +1584,23 @@ class Renderer {
 	}
 
-	renderObject( object, scene, camera, geometry, material, group, lightsNode, passId = null ) {
+	renderObject( object, scene, camera, geometry, material, group, lightsNode, clippingContext = null, passId = null ) {
```
`LensflareMesh` directly calls `renderObject()` and can't access `clippingContext` meaning it can't pass it to `renderObject()`. I've fixed the resulting runtime error by defining a default parameter.

If we add `clippingContext` to `onBeforeRender()`, we could ensure lensflares support clipping as well. Alternatively, we could also use the "current clipping context" for nested renderings if it isn't overwritten with a parameter when calling `renderObject()`.

Mugen87 committed:
Update three.webgpu.js

Revert changes to build files.
